cmd/replay/espresso_validation.go

```diff
@@ -0,0 +1,40 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/offchainlabs/nitro/arbos"
+	"github.com/offchainlabs/nitro/arbos/arbostypes"
+	"github.com/offchainlabs/nitro/wavmio"
+)
+
+// / Handle Espresso Pre Conditions
+// /
+// / Description: This function takes in the most recently recieved message (of type arbostypes.MessageWithMetadata),
+// /	 			  and a boolean from the ChainConfig
+// / 			 and uses the parameters to perform all checks gating the modified STF logic in the arbitrum validator.
+// /
+// / Return: A boolean representing the result of a logical and of all checks (that don't result in panics) gating espressos STF logic.
+// /
+// / Panics: This function will panic if the message type is not an espresso message, but the HotShot height is non zero
+// / 		as this is an invalid state for the STF to reside in.
+// /
+func handleEspressoPreConditions(message *arbostypes.MessageWithMetadata, isEnabled bool) bool {
+	// calculate and cache all values needed to determine if the preconditions are met to enter the Espresso STF logic
+	isNonEspressoMessage := arbos.IsL2NonEspressoMsg(message.Message)
+	isEspressoMessage := !isNonEspressoMessage
+	hotShotHeight := wavmio.GetEspressoHeight()
+	validatingEspressoLivenessFailure := isNonEspressoMessage && (isEnabled || hotShotHeight != 0)
```

Comment:
`wavmio.GetEspressoHeight` will return the current L2 hotshot height. Not every messages in the espresso mode can be called `EspressoMessage`. EspressoMessage is the `L2Message` (which is for creating blocks and consumes 1 hotshot block) in the espresso mode.
There are some other messages and I believe they also have the hotshot height which is equal to the last EspressoMessage

Reply:
I'm a little confused what you mean by "Not all messages in the espresso mode can be called `EspressoMessage`"
I'm aware that we want to allow other messages through when the hotshot height is zero, but if `IsL2NonEspressoMsg` isn't enough to determine the set of messages we want to allow through, what else do I need to look at?

Do we not need to be regularly querying the VM to get the height? It does seem like querying the VM every time could add overhead to the process. Should I just be looking for the height in whatever messages we have access to?

Reply:
https://github.com/EspressoSystems/nitro-espresso-integration/blob/integration/arbos/parse_l2.go#L506
```rs
func IsEspressoMsg(msg *arbostypes.L1IncomingMessage) bool {
	return msg.Header.Kind == arbostypes.L1MessageType_L2Message &&
		msg.L2msg[0] == L2MessageKind_EspressoTx
}

func IsL2NonEspressoMsg(msg *arbostypes.L1IncomingMessage) bool {
	return msg.Header.Kind == arbostypes.L1MessageType_L2Message &&
		msg.L2msg[0] != L2MessageKind_EspressoTx
}
```
The code below is not the same as the condition it was.
```
isNonEspressoMessage := arbos.IsL2NonEspressoMsg(message.Message)
validatingEspressoLivenessFailure := isNonEspressoMessage && (isEnabled || hotShotHeight != 0)
```

Reply:
```
validatingEspressoLivenessFailure := arbos.IsL2NonEspressoMsg(message.Message) && chainConfig.ArbitrumChainParams.EnableEspresso
```

Reply:
The reason that the condition isn't the same is to add the check referenced in the ticket such that the validator panics when the incoming message is not an espresso message, and the hotshot height is non zero.

Reply:
```rs
	isNonEspressoMessage := arbos.IsL2NonEspressoMsg(message.Message)
	isEspressoMessage := !isNonEspressoMessage
```
is not correct. Sorry for the misleading code or interpretion.

Actually this is how we tell if the message is `EspressoMessage` or not:
```rs
func IsEspressoMsg(msg *arbostypes.L1IncomingMessage) bool {
    return msg.Header.Kind == arbostypes.L1MessageType_L2Message && msg.L2msg[0] == L2MessageKind_EspressoTx }
```
and it is not just `!isL2NonEspressoMessage()`.

There are many kinds of messages in the nitro system. The most important one that we need to pay attention to is the `L2Message` with the specific l2message type `EspressoTx` (that is what we called `EspressoMessage`). If the hotshot is good, the l2 messages will have the one type. `EspressoTx`.

When checking liveness failure, we check:
- if nitro are receiving the l2 message that is not `EspressoTx`
- if `espresso` is enabled.